pkg/controller/kubemon/reconciler.go

```diff
@@ -0,0 +1,187 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	Name                   = "kubernetes-monitoring"
+	annotationTemplateHash = "internal.activegate.dynatrace.com/template-hash"
+)
+
+type Reconciler struct {
+	client.Client
+	scheme    *runtime.Scheme
+	dtc       dtclient.Client
+	log       logr.Logger
+	token     *corev1.Secret
+	instance  *v1alpha1.DynaKube
+	apiReader client.Reader
+}
+
+func NewReconciler(clt client.Client, apiReader client.Reader, scheme *runtime.Scheme, dtc dtclient.Client, log logr.Logger, token *corev1.Secret, instance *v1alpha1.DynaKube) *Reconciler {
+	return &Reconciler{
+		Client:    clt,
+		apiReader: apiReader,
+		scheme:    scheme,
+		dtc:       dtc,
+		log:       log,
+		token:     token,
+		instance:  instance,
+	}
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	result, err := dtversion.
+		NewReconciler(r, r.log, r.instance, BuildLabelsFromInstance(r.instance)).
+		Reconcile(request)
+	if err != nil {
+		return result, err
+	}
+
+	result, err = dtpullsecret.
+		NewReconciler(r, r.apiReader, r.scheme, r.instance, r.dtc, r.log, r.token, r.instance.Spec.KubernetesMonitoringSpec.Image).
+		Reconcile(request)
+	if err != nil {
+		r.log.Error(err, "could not reconcile Dynatrace pull secret")
+		return result, err
+	}
+
+	if r.instance.Spec.KubernetesMonitoringSpec.CustomProperties != nil {
+		result, err = customproperties.
+			NewReconciler(r, r.instance, r.log, Name, *r.instance.Spec.KubernetesMonitoringSpec.CustomProperties, r.scheme).
+			Reconcile(request)
+		if err != nil {
+			r.log.Error(err, "could not reconcile custom properties")
+			return result, err
+		}
+	}
+
+	err = r.manageStatefulSet(r.instance)
+	if err != nil {
+		r.log.Error(err, "could not reconcile stateful set")
+		return reconcile.Result{}, err
+	}
+
+	if !r.instance.Spec.KubernetesMonitoringSpec.DisableActivegateUpdate {
+		result, err = dtpods.
+			NewReconciler(r, r.log, r.instance, BuildLabelsFromInstance(r.instance), buildImage(r.instance)).
+			Reconcile(request)
+		if err != nil {
+			r.log.Error(err, "could not update pods")
+			return result, err
+		}
+	}
+
+	if r.instance.Spec.KubernetesMonitoringSpec.KubernetesAPIEndpoint != "" {
+		id, err := r.addToDashboard()
+		r.handleAddToDashboardResult(id, err, r.log)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) manageStatefulSet(instance *v1alpha1.DynaKube) error {
+	desiredStatefulSet, err := r.buildDesiredStatefulSet(instance)
+	if err != nil {
+		return err
+	}
+
+	if err := controllerutil.SetControllerReference(instance, desiredStatefulSet, r.scheme); err != nil {
+		return err
+	}
+
+	currentStatefulSet, err := r.createStatefulSetIfNotExists(desiredStatefulSet)
+	if err != nil {
+		return err
+	}
+
+	err = r.updateStatefulSetIfOutdated(currentStatefulSet, desiredStatefulSet)
```

Comment:
Hmm, if you just created it, no need to update.

Reply:
This function only updates if the data between current and desired has changed, if the StatefulSet was just created, it won't do anything

Reply:
additionally it doesn't query anything from kube-api to check if it is outdated or not and can be safely called after creation